### PR TITLE
:sparkles: [FEAT-53] Onboarding 관련 엔티티 추가

### PR DESCRIPTION
**Pull request by 5hseok**
## #️⃣ Related Issue

Linked Feature backlog: #53

## 📝 Purpose of PR

Onboarding 엔티티 추가 

## Contents

Onboarding 시 질문 목록, 답변 목록, 사용자의 답변 목록을 모두 서버에서 내려주는 것이 좋다고 판단하여 테이블을 따로 만들었습니다. 
처음 생각에서는 질문 내용과 답변은 정해져있으니 테이블을 많이 늘리지 말고 번호로 관리할까 하다가
혹시 질문이나 답변이 이후에 추가될 경우를 생각하여 서버에서 모든 데이터(온보딩 질문, 답변 리스트)를 온보딩 시 클라이언트에게 내려주는 편이 좋다고 판단했습니다. 

## Checklist

-   [ ] Does commit messages follow the convention?
-   [ ] Are variable names brief but descriptive?
-   [ ] Is the code clean and easy to understand?
-   [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
-   [ ] Does the test successfully test edge and corner cases?

## 💬 Review Requirements (Optional)

Is there anything you can praise about this PR? Start with praise.
Make a review, leaving kind comments and suggesting changes where needed (to resolve the above).

### CODE REVIEW

File: src/main/java/ac/dnd/dodal/domain/user_onboarding/model/UserAnswer.java
Comment (5hseok) [inline src/main/java/ac/dnd/dodal/domain/user_onboarding/model/UserAnswer.java:1]: 사용자가 답변을 선택하여 서버에 저장하는 것(UserAnswer)과, 온보딩 질문과 답변 자체를 추가하는 것(Answer, Question)은 도메인이 다르다고 생각해서 패키지를 나눠봤습니다.

File: src/main/java/ac/dnd/dodal/domain/user_onboarding/model/UserAnswer.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+@Table(name = "user_answers")
+public class UserAnswer extends BaseEntity {
+
+  @Id
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  @Column(name = "user_answer_id", nullable = false)
+  private Long id;
+
+  @ManyToOne(fetch = FetchType.LAZY)
+  @JoinColumns({
```
Comment (5hseok) [inline src/main/java/ac/dnd/dodal/domain/user_onboarding/model/UserAnswer.java:28]: UserAnswer 테이블에서 질문과 답변을 한번에 가져오기 위해 Answer 테이블에서는 복합키를 사용하고 이를 불러오는 방식을 고안했습니다.

File: src/main/java/ac/dnd/dodal/domain/onboarding/enums/QuestionContent.java
Comment (5hseok) [inline src/main/java/ac/dnd/dodal/domain/onboarding/enums/QuestionContent.java:1]: Question과 Answer는 수정되어서는 안되고 쉽게 추가할 항목이 아니라고 생각하여 현재 ENUM에서 가져오는 방식으로 개발자 입장에서 보기 편하게 설계해두었는데, 
Question이나 Answer 항목을 관리자가 추가할 수 있게 만들기 위해서는 ENUM이 아니라 바로 DB에서 Low String을 가져오는 방식이 나을 것 같기도 하구요.

여기서는 관리자 입장에서도 온보딩 데이터를 추가하는 것은 대대적인 업데이트가 아닌 이상 일어나지 않을 것이라고 판단하여 ENUM으로 설정해두긴 했습니다.

File: src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  @Id
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  @Column(name = "user_answer_id", nullable = false)
+  private Long id;
+
+  @ManyToOne(fetch = FetchType.LAZY)
+  @JoinColumns({
+    @JoinColumn(name = "question_id", referencedColumnName = "question_id"),
+    @JoinColumn(name = "answer_id", referencedColumnName = "answer_id")
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java:29]: 둘이 동일한 상황에서 referencedColumnName이 필요한가요?

File: src/main/resources/db/migration/V13__add_table_relating_onboarding.sql [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        ON UPDATE CASCADE,
+    FOREIGN KEY (user_id)
+        REFERENCES users(user_id)
+        ON DELETE CASCADE
+        ON UPDATE CASCADE
+);
+
+INSERT INTO questions (question_id, question_content, order_number, created_at, updated_at)
+VALUES
+    (1, 'INTEREST_GOAL', 1, '2025-02-13 00:00:00', '2025-02-13 00:00:00'),
```
Comment (sueelly) [inline src/main/resources/db/migration/V13__add_table_relating_onboarding.sql:53]: 수정하라는 것은 아니고, default current_timestamp 설정하셔서 question_content, order_number만 넣으셔도 insert 알아서 됩니다!

File: src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  })
+  private Answer answer;
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java:31]: 추후에 온보딩 데이터가 바뀐다 할 때를 고려하면 FK로 연결하기보단 그냥 string으로 저장하는 게 낫지 않을까요?
피드백은 이미 그렇게 저장중이거든요. 
FK로 되어었으면 바뀌었을 때 동일한 내용이 아니라면 아예 쓸모없는 정보들이 되어버리고, 내가 선택하지 않은 값이 가져와지는데
온보딩 데이터가 바뀌어도 유저가 선택했던 데이터가 바뀌는 건 아니니까요~

File: src/main/java/ac/dnd/dodal/domain/onboarding/enums/QuestionContent.java
Comment (sueelly) [reply src/main/java/ac/dnd/dodal/domain/onboarding/enums/QuestionContent.java:1]: 온보딩 데이터와 피드백 데이터 (질문/지표 항목들)들은 변동이 많이 일어날 것 같아요
해당 값이 맞는지에 대한 검증이 되지 않았기 때문에 계속 개선되어갈 것 같다고 말씀하셨었거든요.

다만 enum과 db 둘 다 사용한다면 무엇을 기준으로 사용할지, 둘의 싱크를 어떻게 맞출지는 고민을 해야할 것 같아요.
application에서만 수정이 되거나, db에서만 수정이 되거나 할 수 있으니까요.
당장 온보딩 데이터를 그대로 갖고 계산이 들어가야해서 어플리케이션 레벨에서 해당 데이터를 갖고 있어야하고 enum이 필수적인 상황이긴 한데, 이 경우에 db와의 싱크는 어떻게 맞추는 게 좋을까요?

File: src/main/java/ac/dnd/dodal/domain/onboarding/model/Answer.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    private Long answerId;
+
+    @Id
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "question_id", nullable = false)
+    private Question question;
+
+    @Column(name = "answer_content", nullable = false)
+    @Enumerated(EnumType.STRING)
+    private AnswerContent content;
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/domain/onboarding/model/Answer.java:34]: 이렇게 되면 INTEREST_GOAL_1("건강/운동")에서 INTEREST_GOAL_1이 저장될거에요
건강/운동으로 저장되는 게 맞지 않을까요?

File: src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java (hunk 2 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java:29]: @JoinColumns를 사용할 때는 이렇게 사용해주는 편이 이슈를 줄이기에 좋다고 하더라구요!

https://velog.io/@dghff/JPA-%EC%8B%9D%EB%B3%84%EA%B4%80%EA%B3%84%EC%9D%B8-%EB%B3%B5%ED%95%A9%ED%82%A4-%EB%A7%A4%ED%95%91

Comment (sueelly) [reply src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java:29]: 해당되는 이슈가 테이블 생성 시의 FK 순서에 대한 것이라면 저희는 ddl-auto가 아니라 직접 테이블을 정의해서 상관없는 이슈일 것 같아요!
결국 남은 것은 pk가 아닌 값을 fk로 사용할 때 사용하는 용도인데, 현재는 그것도 아닌 듯 합니다!

File: src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java (hunk 4 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java:31]: 온보딩 데이터는 피드백 텍스트나 유형에 영향을 미치는 Base한 값이라고 생각했는데, FK가 아니라 String으로 저장하면 수정사항이 생겼을 때 사용자마다 다른 값을 가지게 되지 않을까요?

온보딩을 마친 사용자가 이미 있다고 가정했을 때 기존 온보딩 데이터를 수정한다면 기존 데이터를 삭제하거나 수정하는 것이 아니라 새로 온보딩 데이터를 DB에 추가하고 새로운 사용자들은 추가한 데이터를 포함해서 온보딩을 진행할 수 있게 하는 것이 맞는 것 같아서요. 물론 이렇게 하면 Answer나 Question 테이블에 nowUse 같은 컬럼으로 현재 서비스에서 제공되고 있는 값인지를 표시하긴 해야겠지만요.

이런 상황에서 String으로 저장한다고 하면 Answer 데이터가 수정되었을 때, Answer 테이블에 있는 값들과 UserAnswer(사용자들이 골랐던 값들)의 값들이 달라지게 될 것 같은데, 이것도 어색하지 않나요?

File: src/main/java/ac/dnd/dodal/domain/onboarding/model/Answer.java (hunk 5 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/domain/onboarding/model/Answer.java:34]: 이렇게 했던 게 사실 DB와 서버의 데이터 싱크를 맞추려고 생각했던 방법이긴 한데, 근본적인 해결책은 아닌 것 같아서 조금 어색한가요?

DB에서 새롭게 추가된 INTEREST_GOAL_5이라는 텍스트를 가져왔는데, 해당하는 ENUM 값이 없으면 아직 서버에 반영되지 않았다는 의미니까 예외가 발생하거나 하는 등으로 처리하기 쉽다고 생각하긴 했어요

File: src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java (hunk 2 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java:29]: Answer 테이블을 정의할 때 answer_id가 AUTO로 생성되는 id값인데, MySQL에서는 자동으로 생성되는 column은 테이블 가장 왼쪽에 위치해야 한다고 해서 answer_id가 먼저 정의되고 question_id가 정의되는 상황입니다.

`CREATE TABLE IF NOT EXISTS answers
(
    answer_id   BIGINT      NOT NULL AUTO_INCREMENT,
    question_id BIGINT      NOT NULL,
    answer_content     VARCHAR(50) NOT NULL,`
    
이러면 순서상 @JoinColumns 내에서 answer_id를 먼저 적고 question_id를 후에 적어야 하는데, referencedColumnName을 붙이는 것과 @JoinColumn 순서를 바꾸는 것 중에서는 어떤 방법이 더 나을까요?

Comment (sueelly) [reply src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java:29]: 제가 이슈를 잘못이해했네요. 위 글에서는 table generator가 따로 정의되어있어 테이블에 컬럼이 정의되는 순서 (insert x, ddl o)로 이해했어요.
삽입되는 순서 자체가 문제였군요. 그렇다면 referenced가 맞는 것 같습니다.
실제 javadoc에도 그렇게 적혀있다고 하네요.
https://medium.com/@SlackBeck/jpa-joincolumns-%EC%82%AC%EC%9A%A9%EC%8B%9C-%EC%A3%BC%EC%9D%98-%EC%82%AC%ED%95%AD-7bc22b98ed9b

File: src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java (hunk 4 above)
Comment (sueelly) [reply src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java:31]: Answer 데이터가 수정되는 방식이라면, 사용자가 선택하지 않은 값이 그대로 사용될 수 있으니까요. 
사용자마다 다른 값을 가지는 게 적절하지 않을까요?

말씀하신대로 수정 없이 새롭게 추가만 하는 형식이면 id로 진행해도 될 것 같습니다.
다만 예를 들어 현재 시간 관리가 소요될 시간 산정과 같은 식으로 '변경'된다고 생각하면, '시간 관리'로 기록되는 게 사용자가 '선택'한 값을 저장한다는 측면에서 더 적절할 것 같아요.
그리고 삭제 없이, 수정도 되지 않는다면 더더욱 fk가 필요한가 싶기도 하고요. 이거는 또하나의 제약사항이니까요

File: src/main/java/ac/dnd/dodal/domain/onboarding/model/Answer.java (hunk 5 above)
Comment (sueelly) [reply src/main/java/ac/dnd/dodal/domain/onboarding/model/Answer.java:34]: db에 저장되는 게 그러면 종류와 개수밖에 없고, 실질적인 내용은 어플리케이션에서 저장한다는 건데, 이 둘을 나눠야할 필요성을 잘 모르겠어요.

db와의 싱크를 맞추는 건 예외로 맞추는 게 아니라 실제로 추가/삭제하는 로직을 구현하여 거기서 맞춰야할 것 같아요. 근데 enum 자체가 상수값이라 이게 어떻게 가능할지.. 모르겠네요
db에 저장하는 이유는 서버의 코드를 변경하지 않고 값을 추가/삭제/변경할 수 있기 때문이고, 서버에 저장하는 이유는 코드를 수정하는 것으로 변경이 용이하기 때문이라고 생각하는데,
어드민을 생각하여 db에 저장하기엔 직접 코드 (enum)도 수정을 해야하고, 코드만 쓰기엔 또 db를 필수적으로 거쳐야하는 게 좀 어색하게 느껴지는 것 같아요

File: src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java (hunk 4 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java:31]: 음 그렇게 되려나요... 확실히 제약 조건이긴 하니까 fk가 아니라 String으로 해도 될 것 같네요!

 근데 다른 얘기긴 하지만 사용자마다 데이터가 너무 달라지면 온보딩 데이터랑 피드백 선택지 조합해서 Tip 출력할 때 불편하지 않나요?

Comment (sueelly) [reply src/main/java/ac/dnd/dodal/domain/user/model/UserAnswer.java:31]: 무조건 함께 버전업해야합니다!!
애초에 같은 문자열인지 아닌지로 판별할 예정이라...

아 그러면 혹시 user나 user_answers에 버전 하나만 추가해주실 수 있을까요?

File: src/main/java/ac/dnd/dodal/domain/onboarding/model/Answer.java (hunk 5 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/domain/onboarding/model/Answer.java:35]: Converter를 두어 수정했습니다!

**Review by sueelly — APPROVED** [review]
